Libraries/libutil/Sources/Options.cpp

```diff
@@ -31,42 +34,93 @@ NextString(std::string *result, std::vector<std::string> const &args, std::vecto
     }
 }
 
+template<>
 std::pair<bool, std::string> Options::
-NextInt(int *result, std::vector<std::string> const &args, std::vector<std::string>::const_iterator *it, bool allowDuplicate)
+Next<int>(ext::optional<int> *result, std::vector<std::string> const &args, std::vector<std::string>::const_iterator *it, bool allowDuplicate)
 {
-    std::string str = std::to_string(*result);
-    std::pair<bool, std::string> success = NextString(&str, args, it, allowDuplicate);
-    if (success.first) {
-        *result = std::atoi(str.c_str());
+    std::string const &arg = **it;
+    if (++*it != args.end()) {
+        std::string const &value = **it;
+
+        if (!*result || allowDuplicate) {
+            *result = std::atoi(value.c_str());
+            return std::make_pair(true, std::string());
+        } else {
+            return std::make_pair(false, "duplicate argument " + arg);
+        }
+    } else {
+        return std::make_pair(false, "missing argument value for argument " + arg);
     }
-    return success;
 }
 
+template<>
 std::pair<bool, std::string> Options::
-NextBool(bool *result, std::vector<std::string> const &args, std::vector<std::string>::const_iterator *it, bool allowDuplicate)
+Next<bool>(ext::optional<bool> *result, std::vector<std::string> const &args, std::vector<std::string>::const_iterator *it, bool allowDuplicate)
 {
-    std::string str = *result ? "YES" : "NO";
-    std::pair<bool, std::string> success = NextString(&str, args, it, allowDuplicate);
-    if (success.first) {
-        if (str == "YES") {
-            *result = true;
-        } else if (str == "NO") {
-            *result = false;
+    std::string const &arg = **it;
+    if (++*it != args.end()) {
+        std::string const &value = **it;
+
+        if (!*result || allowDuplicate) {
+            if (value == "YES") {
```

Comment:
This looks like "no" and "yes" will fail. Is it really expected to be case sensitive?